File: app/services/position_match/sources.py
"""Candidate sources and mask-family source registry."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, Optional

from app.domain.lexicon.ranking import literal_priority_sort_key
from app.models.word import Word
from app.services.position_match.spec import CandidateSource, MatchSpec, SlotConstraint
from app.services.word_db_filters import apply_code_filter, length_filter
from app.services.position_match.mask_adapter import (
    mask_char_glob_pattern,
    mask_fixed_literal_prefix,
    matches_mask_literal_chars,
    required_codes_from_spec,
)
from app.services.word_serializer import get_word_jyutping, get_word_sort_code, get_word_text
from app.utils.jyutping_codec import get_code_variants
from app.utils.word_cache import (
    get_mask_index_candidates,
    get_phoneme_index_candidates,
    get_words_for_length,
    is_word_cache_ready,
)
from app.services._generated.candidate_source_policy import CANDIDATE_FALLBACK_LIMIT


@dataclass
class LengthMaskCandidateSource:
    """Cache-first 長度桶 + mask literal 預過濾（碼字 tail／字面參考）。"""

    db: Any
    mask: str

    def get_candidates(
        self,
        length: int,
        *,
        code: Optional[str] = None,
        mode: str = "m1",
    ) -> tuple[list[Any], bool]:
        return get_length_candidates(self.db, length, self.mask)


@dataclass
class RhymeAnchorCandidateSource:
    """韻／聲錨：cache-ready 時直接走音素倒排索引，跳過全桶掃描。"""

    db: Any
    mask: str
    anchor_pos: int
    anchor: str
    constraint: str

    def get_candidates(
        self,
        length: int,
        *,
        code: Optional[str] = None,
        mode: str = "m1",
    ) -> tuple[list[Any], bool]:
        return get_rhyme_anchor_length_candidates(
            self.db,
            length,
            self.mask,
            self.anchor_pos,
            self.anchor,
            self.constraint,
        )


@dataclass
class LengthCodeCandidateSource:
    """Cache-first 長度桶 + 可選 0243 碼過濾（hybrid 等）。"""

    db: Any
    code: Optional[str] = None
    mode: str = "m1"
    fallback_limit: Optional[int] = CANDIDATE_FALLBACK_LIMIT

    def get_candidates(
        self,
        length: int,
        *,
        code: Optional[str] = None,
        mode: str = "m1",
    ) -> tuple[list[Any], bool]:
        effective_code = code if code is not None else self.code
        effective_mode = mode or self.mode
        return get_candidates_for_length(
            self.db,
            length,
            code=effective_code,
            mode=effective_mode,
            fallback_limit=self.fallback_limit,
        )


@dataclass
class MaskWildcardCandidateSource:
    """缺字查詢：cache literal 預過濾或 DB GLOB + 可選 code filter。"""

    db: Any
    mask: str
    mode: str = "m1"
    query_code: Optional[str] = None
    required_codes: Optional[list[Optional[str]]] = None

    def get_candidates(
        self,
        length: int,
        *,
        code: Optional[str] = None,
        mode: str = "m1",
    ) -> tuple[list[Any], bool]:
        effective_mode = mode or self.mode
        effective_code = code if code is not None else self.query_code
        required_codes = self.required_codes or [None] * len(self.mask)

        indexed = get_mask_index_candidates(length, self.mask)
        if indexed is not None:
            candidates = indexed
        else:
            candidates = get_words_for_length(length)
        if candidates:
            return [
                w for w in candidates
                if matches_mask_literal_chars(get_word_text(w), self.mask)
            ], True

        glob_pat = mask_char_glob_pattern(self.mask)
        query = self.db.query(Word).filter(
            length_filter(length),
            Word.char.op("GLOB")(glob_pat),
        )
        prefix = mask_fixed_literal_prefix(self.mask)
        if prefix:
            query = query.filter(Word.char.like(f"{prefix}%"))
        code_filter = "".join(required_codes) if all(req is not None for req in required_codes) else None
        if code_filter:
            query = apply_code_filter(query, code_filter, effective_mode)
        elif effective_code:
            query = apply_code_filter(query, effective_code, effective_mode)
        # Cold DB: never materialize full length bucket (Portable hang on mask family)
        return (
            query.order_by(Word.char, Word.jyutping).limit(CANDIDATE_FALLBACK_LIMIT).all(),
            False,
        )


def get_length_candidates(
    db,
    width: int,
    mask: str,
    *,
    fallback_limit: Optional[int] = CANDIDATE_FALLBACK_LIMIT,
):
    """
    取得指定長度的候選詞，並對 cache 命中者先做 mask literal 預過濾。
    用於 rhyme-anchor、code-tail、at-tail 等需要 mask 的情境。
    fallback_limit=None → 語意完整候選宇宙（唔截斷）。
    """
    indexed = get_mask_index_candidates(width, mask)
    if indexed is not None:
        candidates = indexed
    else:
        candidates = get_words_for_length(width)
    if candidates:
        return [w for w in candidates if matches_mask_literal_chars(get_word_text(w), mask)], True
    glob_pat = mask_char_glob_pattern(mask)
    query = db.query(Word).filter(
        length_filter(width),
        Word.char.op("GLOB")(glob_pat),
    )
    prefix = mask_fixed_literal_prefix(mask)
    if prefix:
        query = query.filter(Word.char.like(f"{prefix}%"))
    query = query.order_by(Word.char, Word.jyutping)
    if fallback_limit is not None:
        query = query.limit(fallback_limit)
    return query.all(), False


def get_rhyme_anchor_length_candidates(
    db,
    width: int,
    mask: str,
    anchor_pos: int,
    anchor: str,
    constraint: str,
) -> tuple[list[Any], bool]:
    """韻／聲錨候選：音素索引優先；冷路徑必須語意完整候選宇宙（唔 LIMIT 再濾韻）。"""
    if is_word_cache_ready():
        rows = get_phoneme_index_candidates(width, anchor_pos, anchor, constraint, db)
        narrowed = [
            w for w in rows
            if matches_mask_literal_chars(get_word_text(w), mask)
        ]
        return narrowed, True
    # Cold: mask GLOB without CANDIDATE_FALLBACK_LIMIT — phoneme filter runs in apply_match_spec.
    # (Repro: 就= with cache cold missed 后／就 when LIMIT 2000 before rhyme filter.)
    return get_length_candidates(db, width, mask, fallback_limit=None)


def get_candidates_for_length(
    db: Any,
    length: int,
    *,
    code: Optional[str] = None,
    mode: str = "m1",
    fallback_limit: Optional[int] = CANDIDATE_FALLBACK_LIMIT,
):
    """
    通用長度候選取得（無 mask 預過濾）。
    用於 hybrid 等情境。fallback_limit=None 表示唔截斷（碼夾 hybrid 需全桶）。
    """
    candidates = get_words_for_length(length)
    if candidates:
        return candidates, True
    query = db.query(Word).filter(length_filter(length))
    if code:
        query = apply_code_filter(query, code, mode)
    query = query.order_by(Word.char, Word.jyutping)
    if fallback_limit is not None:
        query = query.limit(fallback_limit)
    return query.all(), False


def _phoneme_anchor_slot(spec: MatchSpec) -> Optional[SlotConstraint]:
    for slot in spec.slots:
        if slot.kind in ("final_anchor", "initial_anchor"):
            return slot
    return None


def _compound_rhyme_char(spec: MatchSpec) -> Optional[str]:
    for slot in spec.slots:
        if slot.kind == "final_anchor" and isinstance(slot.value, str):
            return slot.value
    return None


@dataclass
class CompoundCandidateSource:
    """近義／反義／連接詞複合：字面容許集 + char IN；缺庫字面記憶體合成（ADR-0054）。"""

    db: Any
    compounds: frozenset[str]
    expected_length: int = 2
    allow_transient: bool = False

    def get_candidates(
        self,
        length: int,
        *,
        code: Optional[str] = None,
        mode: str = "m1",
    ) -> tuple[list[Any], bool]:
        if length != self.expected_length or not self.compounds:
            return [], True

        rows: list[Any] = []
        from_cache = False
        if is_word_cache_ready():
            rows = [
                w for w in get_words_for_length(self.expected_length)
                if get_word_text(w) in self.compounds
            ]
            from_cache = True
        if not rows:
            query = self.db.query(Word).filter(
                Word.char.in_(list(self.compounds)),
                length_filter(self.expected_length),
            )
            rows = query.order_by(Word.char, Word.code, Word.jyutping).all()
            from_cache = False

        if self.allow_transient:
            have = {get_word_text(w) for w in rows}
            missing = [ch for ch in self.compounds if ch not in have]
            if missing:
                from app.domain.relations.compound_connect import compose_transient_word

                for ch in missing:
                    transient = compose_transient_word(ch)
                    if transient is not None:
                        rows.append(transient)

        return rows, from_cache


CompoundSynCandidateSource = CompoundCandidateSource
CompoundAntCandidateSource = CompoundCandidateSource


def _resolve_mask_family_source(
    spec: MatchSpec,
    db: Any,
    mode: str,
    query_code: Optional[str],
) -> tuple[Optional[CandidateSource], Optional[Callable]]:
    """由 MatchSpec 形狀選擇候選來源（registry，不依 ParsedQuery）。"""
    from app.domain.lexicon.ranking import search_result_sort_key

    connective = spec.extra.get("connective")
    if spec.compound_kind == "doubled_syllable":
        rhyme_char = _compound_rhyme_char(spec)
        from app.domain.relations.compound_doubled_syllable import search_doubled_syllable

        tiers = search_doubled_syllable(db, rhyme_char=rhyme_char, width=spec.width)
        if not tiers:
            return None, None
        source = CompoundCandidateSource(
            db, frozenset(tiers.keys()), expected_length=spec.width
        )
        sort_key = lambda w: (tiers.get(get_word_text(w), 99), search_result_sort_key(w))
        return source, sort_key

    if spec.compound_kind == "syn":
        rhyme_char = _compound_rhyme_char(spec)
        if connective and spec.width == 3:
            from app.domain.relations.compound_connect import search_connective_compound

            tiers = search_connective_compound(
                db,
                compound_kind="syn",
                connective=connective,
                rhyme_char=rhyme_char,
            )
            if not tiers:
                return None, None
            source = CompoundCandidateSource(
                db,
                frozenset(tiers.keys()),
                expected_length=3,
                allow_transient=True,
            )
            sort_key = lambda w: (tiers.get(get_word_text(w), 99), search_result_sort_key(w))
            return source, sort_key

        from app.domain.relations.compound_syn import search_compound_syn

        tiers = search_compound_syn(db, rhyme_char=rhyme_char, width=spec.width)
        if not tiers:
            return None, None
        source = CompoundCandidateSource(db, frozenset(tiers.keys()))
        sort_key = lambda w: (tiers.get(get_word_text(w), 99), search_result_sort_key(w))
        return source, sort_key

    if spec.compound_kind == "ant":
        rhyme_char = _compound_rhyme_char(spec)
        if connective and spec.width == 3:
            from app.domain.relations.compound_connect import search_connective_compound

            tiers = search_connective_compound(
                db,
                compound_kind="ant",
                connective=connective,
                rhyme_char=rhyme_char,
            )
            if not tiers:
                return None, None
            source = CompoundCandidateSource(
                db,
                frozenset(tiers.keys()),
                expected_length=3,
                allow_transient=True,
            )
            sort_key = lambda w: (tiers.get(get_word_text(w), 99), search_result_sort_key(w))
            return source, sort_key

        from app.domain.relations.compound_ant import search_compound_ant

        tiers = search_compound_ant(db, rhyme_char=rhyme_char, width=spec.width)
        if not tiers:
            return None, None
        source = CompoundCandidateSource(db, frozenset(tiers.keys()))
        sort_key = lambda w: (tiers.get(get_word_text(w), 99), search_result_sort_key(w))
        return source, sort_key

    if spec.literal_priority and spec.mask:
        from app.services.position_match.mask_adapter import dense_code_from_spec

        effective_code = query_code or dense_code_from_spec(spec)
        source = MaskWildcardCandidateSource(
            db,
            spec.mask,
            mode=mode,
            query_code=effective_code,
            required_codes=required_codes_from_spec(spec),
        )
        literal_positions = spec.extra.get("literal_positions", [])
        sort_key = lambda w: literal_priority_sort_key(w, literal_positions)
        return source, sort_key

    anchor = _phoneme_anchor_slot(spec)
    if anchor and spec.mask and not spec.literal_priority:
        constraint = "final" if anchor.kind == "final_anchor" else "initial"
        source = RhymeAnchorCandidateSource(
            db,
            spec.mask,
            anchor.pos,
            anchor.value or "",
            constraint,
        )
        return source, None

    # TS parity (specNeedsFullLengthBucket): letter slots need full bucket —
    # LengthMaskCandidateSource(??) + 2000 LIMIT truncates before letter filter
    # (repro: 43$獅 → 43si misses 舞獅).
    has_letter_slots = any(
        s.kind in ("rhyme_letters", "syllable_letters", "initial_letters")
        for s in spec.slots
    )
    if has_letter_slots:
        from app.services.position_match.mask_adapter import dense_code_from_spec

        effective_code = query_code or dense_code_from_spec(spec)
        return (
            LengthCodeCandidateSource(
                db,
                code=effective_code,
                mode=mode,
                fallback_limit=None,
            ),
            None,
        )

    if spec.mask:
        from app.services.position_match.mask_adapter import dense_code_from_spec

        # Full-width dense code: load complete code bucket (no LIMIT 2000 + ORDER BY char).
        # Repro: workbench 貪婪 same_tone → code 30; 金錢 high essay but late alpha → dropped.
        effective_code = query_code or dense_code_from_spec(spec)
        if (
            effective_code
            and len(effective_code) == spec.width
            and effective_code.isdigit()
        ):
            return (
                LengthCodeCandidateSource(
                    db,
                    code=effective_code,
                    mode=mode,
                    fallback_limit=None,
                ),
                None,
            )
        return LengthMaskCandidateSource(db, spec.mask), None

    return None, None

File: client/src/db/position-match/sources.ts
/**
 * Candidate sources — port of app/services/position_match/sources.py (MF-3 + MF-5 F5)
 */
import { fetchCompoundWordRows, type CompoundSearchSpec } from '../compound.ts';
import { getCodeVariants } from '../code-variants.ts';
import { queryRows } from '../database-backend.ts';
import type { Database } from '../sqljs.ts';
import {
  maskCharGlobPattern,
  maskFixedLiteralPrefix,
  matchesMaskLiteralChars,
} from './mask-adapter.ts';
import type { CandidateSource, MatchSpec } from './spec.ts';
import {
  CANDIDATE_FALLBACK_LIMIT,
  lengthBucketNeedsUnlimited,
} from './candidate-policy.ts';

export type WordRow = Record<string, unknown>;
export { CANDIDATE_FALLBACK_LIMIT, lengthBucketNeedsUnlimited } from './candidate-policy.ts';

export function wordMatchesWidth(row: WordRow, width: number): boolean {
  const stored = Number(row.length ?? 0);
  if (stored > 0) {
    return stored === width;
  }
  return String(row.char ?? '').length === width;
}

export type GetCandidatesOptions = {
  code?: string | null;
  mode?: string;
  fallbackLimit?: number;
  /** ponytail: jyutping letter anchors need full length bucket (desktop word_cache parity) */
  unlimited?: boolean;
};

/**
 * 通用長度候選取得（無 mask 預過濾）— port of get_candidates_for_length.
 * ponytail: PWA 無 word_cache，一律走 DB；第二返回值恒 false。
 */
export async function getCandidatesForLength(
  db: Database,
  length: number,
  options: GetCandidatesOptions = {},
): Promise<[WordRow[], boolean]> {
  const mode = options.mode === 'm2' || options.mode === '02493' ? 'm2' : 'm1';
  const limit = options.fallbackLimit ?? CANDIDATE_FALLBACK_LIMIT;
  const unlimited = lengthBucketNeedsUnlimited(options);
  const code = options.code ?? null;

  let sql = `
    SELECT char, jyutping, code, initials, finals, length
    FROM words
    WHERE (
      length = ?
      OR ((length IS NULL OR length = 0) AND length(char) = ?)
    )
  `;
  const params: Array<string | number> = [length, length];

  if (code) {
    const variants = getCodeVariants(code, mode);
    if (variants.length) {
      sql += ` AND code IN (${variants.map(() => '?').join(', ')})`;
      params.push(...variants);
    }
  }

  sql += ' ORDER BY char, jyutping';
  if (!unlimited) {
    sql += ' LIMIT ?';
    params.push(limit);
  }

  const resultRows = await queryRows(db, sql, params);
  const rows: WordRow[] = [];
  for (const row of resultRows) {
    if (wordMatchesWidth(row, length)) {
      rows.push(row);
    }
  }
  return [rows, false];
}

/** Port of sources.get_length_candidates — mask GLOB，唔受 2000 上限 */
export async function getLengthMaskCandidates(
  db: Database,
  length: number,
  mask: string,
  options: { code?: string | null; mode?: string } = {},
): Promise<[WordRow[], boolean]> {
  const mode = options.mode === 'm2' || options.mode === '02493' ? 'm2' : 'm1';
  const globPat = maskCharGlobPattern(mask);
  const prefix = maskFixedLiteralPrefix(mask);
  let sql = `
    SELECT char, jyutping, code, initials, finals, length
    FROM words
    WHERE (
      length = ?
      OR ((length IS NULL OR length = 0) AND length(char) = ?)
    )
    AND char GLOB ?
  `;
  const params: Array<string | number> = [length, length, globPat];
  if (prefix) {
    sql += ' AND char LIKE ?';
    params.push(`${prefix}%`);
  }
  const code = options.code ?? null;
  if (code) {
    const variants = getCodeVariants(code, mode);
    if (variants.length) {
      sql += ` AND code IN (${variants.map(() => '?').join(', ')})`;
      params.push(...variants);
    }
  }
  sql += ' ORDER BY char, jyutping';

  const resultRows = await queryRows(db, sql, params);
  const rows: WordRow[] = [];
  for (const row of resultRows) {
    const text = String(row.char ?? '');
    if (wordMatchesWidth(row, length) && matchesMaskLiteralChars(text, mask)) {
      rows.push(row);
    }
  }
  return [rows, false];
}

function maskWithLiteralAt(width: number, pos: number, ch: string): string {
  return Array.from({ length: width }, (_, i) => (i === pos ? ch : '?')).join('');
}

/** 錨字在固定格的所有 width 詞（partial mask slot options） */
export async function getCandidatesWithLiteralAt(
  db: Database,
  width: number,
  pos: number,
  ch: string,
): Promise<WordRow[]> {
  return (await getLengthMaskCandidates(db, width, maskWithLiteralAt(width, pos, ch)))[0];
}

/** Port of LengthCodeCandidateSource */
export class LengthCodeCandidateSource implements CandidateSource {
  constructor(
    private readonly db: Database,
    private readonly code: string | null = null,
    private readonly mode = 'm1',
    private readonly fallbackLimit = CANDIDATE_FALLBACK_LIMIT,
  ) {}

  async getCandidates(
    length: number,
    options?: { code?: string | null; mode?: string },
  ): Promise<[unknown[], boolean]> {
    const effectiveCode = options?.code !== undefined ? options.code : this.code;
    const effectiveMode = options?.mode ?? this.mode;
    return getCandidatesForLength(this.db, length, {
      code: effectiveCode,
      mode: effectiveMode,
      fallbackLimit: this.fallbackLimit,
    });
  }
}

/** Port of sources._compound_rhyme_char */
export function compoundRhymeChar(spec: MatchSpec): string | undefined {
  for (const slot of spec.slots ?? []) {
    if (slot.kind === 'final_anchor' && slot.value) {
      return String(slot.value);
    }
  }
  return undefined;
}

export function compoundSearchSpecFromMatchSpec(spec: MatchSpec): CompoundSearchSpec | null {
  if (!spec.compound_kind) {
    return null;
  }
  const connective = spec.extra?.connective;
  return {
    compound_kind: spec.compound_kind,
    width: spec.width,
    code_prefix: spec.code_prefix ?? undefined,
    rhyme_char: compoundRhymeChar(spec),
    connective: typeof connective === 'string' ? connective : undefined,
  };
}

/** Port of CompoundCandidateSource */
export class CompoundCandidateSource implements CandidateSource {
  constructor(
    private readonly db: Database,
    private readonly compounds: Set<string>,
    private readonly expectedLength = 2,
  ) {}

  async getCandidates(
    length: number,
    _options?: { code?: string | null; mode?: string },
  ): Promise<[WordRow[], boolean]> {
    if (length !== this.expectedLength || !this.compounds.size) {
      return [[], false];
    }
    const list = [...this.compounds];
    const placeholders = list.map(() => '?').join(', ');
    const resultRows = await queryRows(this.db, `
      SELECT char, jyutping, code, initials, finals, length
      FROM words
      WHERE char IN (${placeholders})
        AND (
          length = ?
          OR ((length IS NULL OR length = 0) AND length(char) = ?)
      )
      ORDER BY char, jyutping
    `, [...list, this.expectedLength, this.expectedLength]);
    const rows: WordRow[] = [];
    for (const row of resultRows) {
      if (wordMatchesWidth(row, this.expectedLength)) {
        rows.push(row);
      }
    }
    return [rows, false];
  }
}

/** ponytail: MF-5 F5 compound_kind source — reuses compound.ts tier search */
export async function getCompoundCandidatesForSpec(
  spec: MatchSpec,
  db: Database,
  mode: string,
): Promise<WordRow[]> {
  const compoundSpec = compoundSearchSpecFromMatchSpec(spec);
  if (!compoundSpec) {
    return [];
  }
  return fetchCompoundWordRows(db, compoundSpec, mode);
}

/** ponytail: runnable self-check — needs injected Database */
export async function positionMatchSourcesSelfCheck(db: Database): Promise<void> {
  const [width2] = await getCandidatesForLength(db, 2);
  if (!width2.length) {
    throw new Error('positionMatchSourcesSelfCheck: width=2 bucket empty');
  }
  const sampleCode = String(width2[0]!.code ?? '');
  if (!sampleCode) {
    throw new Error('positionMatchSourcesSelfCheck: sample row missing code');
  }
  const [filtered] = await getCandidatesForLength(db, 2, { code: sampleCode });
  if (!filtered.length) {
    throw new Error('positionMatchSourcesSelfCheck: code filter empty');
  }
  const source = new LengthCodeCandidateSource(db, sampleCode, 'm1');
  const [viaSource] = await source.getCandidates(2);
  if (!viaSource.length) {
    throw new Error('positionMatchSourcesSelfCheck: LengthCodeCandidateSource');
  }
}
